src/header/TransferBench.hpp

```diff
@@ -2088,6 +2272,1114 @@ namespace {
+    const char* deviceName = ibv_get_device_name(context->device);
+    ERR_CHECK(GetRoceVersionNumber(deviceName, portNum, *gidIndex, &gidRoceVerNum));    
+    ERR_CHECK(GetRoceVersionNumber(deviceName, portNum, gidIndexCandidate, &gidRoceVerNumCandidate));    
+    if ((gidRoceVerNum != gidRoceVerNumCandidate || !IsValidGid(&gid)) && gidRoceVerNumCandidate == usrRoceVer)
+    {
+      *gidIndex = gidIndexCandidate;
+    }
+  }
+  return ERR_NONE;
+}
+
+static ErrResult SetGidIndex(struct ibv_context* context,
+                             uint8_t const&      portNum,
+                             int const&          gidTblLen,
+                             int const&          roceVersion,
+                             int const&          ipAddressFamily,
+                             int*                gidIndex
+                            )
+{
+  if (*gidIndex >= 0) {
+    return ERR_NONE;
+  }
+  sa_family_t userAddrFamily = (ipAddressFamily == 6)? AF_INET6 : AF_INET;
+
+  int userRoceVersion = roceVersion;
+
+  // TODO: Get address range from user
+  void *prefix = NULL;
+
+  *gidIndex = 0;
+  for (int gidIndexNext = 1; gidIndexNext < gidTblLen; ++gidIndexNext) {
+    ErrResult err = UpdateGidIndex(context, portNum, userAddrFamily,
+                                   prefix, 0, userRoceVersion, gidIndexNext,
+                                   gidIndex
+                                  );
+    if (err.errType != ERR_NONE) {
+      return err;
+    }
+  }
+  return ERR_NONE;
+}
+
+static ErrResult InitDeviceList() 
+{
+  if (deviceList == NULL) 
+  {
+    IBV_PTR_CALL(deviceList, ibv_get_device_list, &RdmaNicCount);
+  }
+  return ERR_NONE;
+}
+
+static ErrResult GetNicCount(int& NicCount)
+{
+  if (deviceList == NULL && RdmaNicCount < 0)
+  {
+    InitDeviceList();
+  }
+  NicCount = RdmaNicCount;
+  return ERR_NONE;
+}
+
+static ErrResult InitRdmaResources(int                   const& deviceID,
+                                   uint8_t               const& port,
+                                   vector<NicResources*> &      resourceMapper
+                                   ) 
+{
+  if (resourceMapper.size() <= deviceID) {
+    resourceMapper.resize(deviceID + 1);
+    resourceMapper[deviceID] = nullptr; 
+  }
+  if (resourceMapper[deviceID] == nullptr) {
+    resourceMapper[deviceID] = new NicResources();
+    auto& rdma = resourceMapper[deviceID];
+    IBV_PTR_CALL(rdma->context, ibv_open_device, deviceList[deviceID]);
+    IBV_PTR_CALL(rdma->protectionDomain, ibv_alloc_pd, rdma->context);
+    IBV_PTR_CALL(rdma->completionQueue, ibv_create_cq, rdma->context, 100, NULL, NULL, 0);
+    IBV_CALL(ibv_query_port, rdma->context, port, &rdma->portAttr);
+    if (rdma->portAttr.state != IBV_PORT_ACTIVE) {      
+      return {ERR_FATAL, "Selected RDMA device %d is down", deviceID};
+    }
+  }
+  return ERR_NONE;
+}
+
+static ErrResult InitRdmaTransferResources(Transfer          const& transfer,
```

Comment:
By convention, we place input arguments before in/out, output arguments.
I'd order this with rdmaOptions first, followed by Transfer, followed by resources.